### ISSUE TITLE
Backtracking resolver error with extra and URL

### ISSUE THREAD

**Issue by AntoineD** [jazzband/pip-tools#1920]
The following works with the legacy resolver:
```console
pip-compile --upgrade-package "package[all]@git+https://<URL here>"
```
but with the backtracking resolver I get
```console
AssertionError: Internal issue: Candidate is not for this requirement package vs package[all]
```

#### Environment Versions

1. OS Type: fedora 38
1. Python version: `$ python -V`: Python 3.9.17
1. pip version: `$ pip --version`: pip 22.3.1
1. pip-tools version: `$ pip-compile --version`: pip-compile, version 7.0.0

#### Steps to replicate

1. `git clone git@gitlab.com:gemseo/dev/gemseo-umdo.git`
2. `cd gemseo-umdo`
3. `tox -e update-deps-test-py39`

#### Expected result

No error.

#### Actual result

```console
AssertionError: Internal issue: Candidate is not for this requirement package vs package[all]
```

**Comment by atugushev** [comment]
Pip has [changed its support](https://pip.pypa.io/en/stable/user_guide/#changes-to-the-pip-dependency-resolver-in-20-3-2020) for constraints in the backtracking resolver. See https://github.com/pypa/pip/issues/8253 for details (TLDR; URLs are no longer supported). Since the upgrade-package essentially acts as a constraint, this is why it won't work with the backtracking resolver.

```
AssertionError: Internal issue: Candidate is not for this requirement gemseo vs gemseo[all]
...
exit 1: pip-compile --upgrade --upgrade-package 'gemseo[all]@git+https://gitlab.com/gemseo/dev/gemseo.git@develop' --extra test -o requirements/test-py39.txt pid=364
```

Try to pass `--upgrade-package="gemseo[all]"` instead.

**Comment by atugushev** [comment]
> I haven't read yet the link you shared (I will, thank you), but when no extra is needed, the backtracking resolver works fine with an URL:


Oh, sorry. I copied wrong info. I mean extras (not URL):

> Constraints cannot have extras (see https://github.com/pypa/pip/issues/6628).

**Comment by atugushev** [comment]
> Do you think this could be addressed by [pypa/pip#12025](https://github.com/pypa/pip/pull/12025)?

It doesn't seem so. Extras won't be allowed in constraints anyway.

**Comment by atugushev** [comment]
I'll close this since it doesn't look like there's any action required regarding pip-tools. Thanks for the issue!
